transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -718,12 +749,53 @@ protected void doDisconnect() throws Exception {
 
     @Override
     protected void doClose() throws Exception {
+        if (isNotEmpty(tempInboundMessages)) {
+            ClosedChannelException exception = null;
+            for (;;) {
+                AbstractMap.SimpleEntry<Object, ChannelPromise> entry = tempInboundMessages.poll();
+                if (entry == null) {
+                    break;
+                }
+                Object value = entry.getKey();
+                if (value != null) {
+                    ReferenceCountUtil.release(value);
+                }
+                ChannelPromise promise = entry.getValue();
+                if (promise != null) {
+                    if (exception == null) {
+                        exception = new ClosedChannelException();
+                    }
+                    promise.tryFailure(exception);
+                }
+            }
+        }
         state = State.CLOSED;
```

Comment:
better update state before promises are notified to ensure we see the correct state in the futures.

Reply:
Fixed.